### PR TITLE
[GO] Asset Endpoint

### LINKED ISSUE #1366: [GO] Asset endpoint complete
## 📖 Description
Currently Asset endpoint only has GET method implemented. We need to implement the other 3.

---

## 🎯 Goal / Outcome
Asset endpoint correctly handles GET, POST, PUT and DELETE
unit tests

---

## ✅ Definition of Done (DoD)
- [ ] Feature is implemented and works as expected
- [ ] Edge cases are handled
- [ ] Tests are added or updated (if applicable)
- [ ] Documentation is updated (if needed)
- [ ] Code is reviewed and approved

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

---
## 🚧 Risks / Open Questions (Optional)

---

### PR DESCRIPTION
**Pull request by nelgd**
## 📖 Description

Adds write support for assets to the Go SDK. This PR introduces three new asset handlers `LoadAsset`, `OverwriteAsset` , and `DeleteAsset`. Business logic decides create-vs-conflict and overwrite-vs-not-found semantics; the three write operations return `204 No Content` on success.

---

## ✅ Definition of Done (DoD)

- [x] LoadAsset, OverwriteAsset, and DeleteAsset handlers implemented and wired through `Server` + `ServerRuntime`
- [x] REST transport dispatches `POST`/`PUT`/`DELETE` on `/st2138-api/v1/{slot}/asset/{oid}` (returns `204` on success)
- [x] `FromAsset` helper added (inverse of `ToAsset`) plus `ReadAssetRequestJSON` request decoder with Content-Type validation
- [x] Edge cases handled (missing/invalid Content-Type, unparseable body, already-exists / not-found, unregistered handler, unsupported method)
- [x] Tests added/updated (`server_test.go`, `rest_transport_test.go`, `json_utils_test.go`, `stub_server_test.go`)
- [x] Documentation updated (`transports/README.md`, `transports/REST.md`)
- [x] Example updated (`oneOfEverything/asset_handler.go`)

---

## 🛠️ Implementation Notes

- `sdks/go/pkg/catena/asset.go` — added `FromAsset` (converts an `Asset` back into a `DataPayload`, the inverse of `ToAsset`)
- `sdks/go/pkg/catena/server.go`:
  - added `EndpointLoadAsset`, `EndpointOverwriteAsset`, `EndpointDeleteAsset` endpoint types (+ `String()` cases)
  - added `LoadAssetHandler`, `OverwriteAssetHandler`, `DeleteAssetHandler` types
  - extended `Server` and `ServerRuntime` interfaces with `Register*`/`Invoke*` methods for the three handlers
  - added handler maps + registration/invocation implementations on `server`
- `sdks/go/pkg/transports/json_utils.go` — added `ReadAssetRequestJSON` (decodes a JSON body into `protos.ExternalObjectPayload`, validates Content-Type is `application/json`)
- `sdks/go/pkg/transports/rest_transport.go`:
  - `handleAssetEndpoint` now dispatches by method (GET/POST/PUT/DELETE)
  - split out `handleGetAsset` and added `handleWriteAsset` (shared by POST/PUT), write ops return `204 No Content`
- `sdks/go/examples/oneOfEverything/asset_handler.go` — registered Load/Overwrite/Delete handlers backed by a `sync.Map`
- Docs: updated `sdks/go/pkg/transports/README.md` and `sdks/go/pkg/transports/REST.md` with the new routes/handlers
- Tests: `sdks/go/pkg/catena/server_test.go`, `sdks/go/pkg/transports/rest_transport_test.go`, `sdks/go/pkg/transports/json_utils_test.go`, `sdks/go/pkg/transports/stub_server_test.go`

---

## 🔗 Related Issues / Links

Closes issue:
- #1366

### CODE REVIEW

File: sdks/go/pkg/catena/asset.go [hunk 1]
```diff
@@ -211,9 +211,17 @@ func dataPayloadFromProto(pdp *protos.DataPayload) (DataPayload, StatusResult) {
 	}
 	switch k := pdp.GetKind().(type) {
 	case *protos.DataPayload_Url:
+		if k.Url == "" {
+			return DataPayload{}, StatusResult{Code: StatusCodeInvalidArgument, Error: "either payload or url must be provided in DataPayload"}
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/asset.go:215]: coverage

File: sdks/go/pkg/catena/server.go [hunk 2]
```diff
@@ -64,6 +64,9 @@ const (
 	EndpointGetParam
 	EndpointSetValue
 	EndpointGetAsset
+	EndpointLoadAsset
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go]: I'd prefer Get/Add/Update/Delete

File: sdks/go/pkg/catena/server.go [hunk 3]
```diff
@@ -778,6 +822,33 @@ func (s *server) InvokeGetAssetHandler(slot uint16, fqoid string, transportConte
 		})
 }
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go:824]: from the spec: these not only require a write scope, but they also require adm/cfg/op (any but mon)

File: sdks/go/pkg/transports/json_utils.go [hunk 4]
```diff
@@ -187,6 +187,38 @@ func WriteProtoJSON(w http.ResponseWriter, msg proto.Message, statusCode int) er
 	return writeErr
 }
 
+// ReadAssetRequestJSON reads and unmarshals a JSON request body into a
+// protos.ExternalObjectPayload (the inverse of MarshalAssetJSON). It validates
+// the Content-Type header and returns an error if it's not application/json.
+func ReadAssetRequestJSON(r *http.Request) (*protos.ExternalObjectPayload, catena.StatusResult) {
```
Comment (abrown-RV) [inline sdks/go/pkg/transports/json_utils.go]: this should just return an error, you can use `fmt.Errorf("Error message")` then in the callsite its just
```go
payload, err := ReadAssetRequestJSON(r)
if err != nil {
	logger.Error("failed to read asset request", "error", err)
	t.writeHTTPStatusResult(w, catena.StatusWithCode(catena.StatusCodeInvalidArgument, err.Error()))
	return
}
```
more go idomatic and we want to overhaul everything to use this style for these internal error returns.

**Review by abrown-RV — CHANGES_REQUESTED** [review]

File: sdks/go/pkg/catena/server.go (hunk 2 above)
Comment (abrown-RV) [reply sdks/go/pkg/catena/server.go]: Actually lets just do CRUD

Comment (nelgd) [reply sdks/go/pkg/catena/server.go]: So keep it as is? Or should I change the names to

`ReadAsset`, `CreateAsset`, `UpdateAsset`, `DeleteAsset`

Comment (abrown-RV) [reply sdks/go/pkg/catena/server.go]: change the names please

File: sdks/go/pkg/catena/handler_context.go [hunk 5]
```diff
@@ -115,6 +115,18 @@ func (ctx HandlerContext) HasAnyReadScope() bool {
 	return slices.ContainsFunc(catenaScopes, ctx.HasReadScope)
 }
 
+// HasAnyWriteScopeExceptMonitor reports whether the caller holds a write scope
+// other than the monitor scope - that is, op, cfg, or adm write. Per ST 2138
+// the asset mutations (LoadAsset/OverwriteAsset/DeleteAsset) accept adm:w,
+// op:w, or cfg:w but never mon:w, so HasAnyWriteScope is too permissive for
+// them. It returns true for every scope when authorization is disabled.
+func (ctx HandlerContext) HasAnyWriteScopeExceptMonitor() bool {
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/handler_context.go]: 2 layer helper function seems overkill, I think take this one out and just have the one in server.go

File: sdks/go/pkg/catena/server.go [hunk 6]
```diff
@@ -246,7 +272,10 @@ type server struct {
 	getValueHandlers       map[uint16]GetValueHandler
 	getParamHandlers       map[uint16]GetParamHandler
 	setValueHandlers       map[uint16]SetValueHandler
-	getAssetHandlers       map[uint16]GetAssetHandler
+	readAssetHandlers       map[uint16]ReadAssetHandler
+	createAssetHandlers      map[uint16]CreateAssetHandler
+	updateAssetHandlers map[uint16]UpdateAssetHandler
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go]: fix formatting

**Review by abrown-RV — APPROVED** [review]
